linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala

```diff
@@ -375,112 +375,152 @@ private class AnalyzerRun(config: CommonPhaseConfig, initial: Boolean,
       infos
     }
 
-    private def lookupClassForLinking(className: ClassName,
-        origin: LoadingClass): Future[LoadingResult] = {
-      ensureLoading(className, syntheticKind = None) match {
-        case loading: LoadingClass => loading.requestLink(origin)
-        case info: ClassInfo       => Future.successful(info)
-      }
-    }
-
     private def ensureLoading(className: ClassName,
         syntheticKind: Option[SyntheticClassKind]): ClassLoadingState = {
-      var loading: LoadingClass = null
+      var loading: LoadingInfos = null
       val state = _classInfos.getOrElseUpdate(className, {
-        loading = new LoadingClass(className)
+        loading = new LoadingInfos
         loading
       })
```

Comment:
Looking at this, I wonder if we need to mark `loading` as `@volatile`, to ensure safe publication of the assignment across threads? Or perhaps an `AtomicReference`?

Reply:
> No, I don't think so. `loading` is only ever written to by the calling thread (at least for any reasonable implementation of `getOrElseUpdate`).

Right, indeed. Sounds good.